documentation/jetty-documentation/src/main/asciidoc/programming-guide/server/websocket/server-websocket-filter.adoc

```diff
@@ -0,0 +1,79 @@
+//
+// ========================================================================
+// Copyright (c) 1995-2021 Mort Bay Consulting Pty Ltd and others.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+[[pg-server-websocket-configure-filter]]
+==== Advanced `WebSocketUpgradeFilter` Configuration
+
+The WebSocket ``ServletContainerInitializer``s, namely `JavaxWebSocketServletContainerInitializer` and `JettyWebSocketServletContainerInitializer`, install the `WebSocketUpgradeFilter` to handle HTTP requests that upgrade to WebSocket.
+
+Typically, the `WebSocketUpgradeFilter` is not present in the `web.xml` configuration, and therefore the WebSocket ``ServletContainerInitializer``s create a new `WebSocketUpgradeFilter` and install it _before_ any other Filter declared in `web.xml`, under the default name of `"org.eclipse.jetty.websocket.servlet.WebSocketUpgradeFilter"` and with path mapping `/*`.
```

Comment:
Once we have described all the bits that are initialized, then we need subsections saying how each bit of machinery can be customized.  The filter being only one of these subsections.  We also need to say how pools etc. can be customized/configured/replaced.

Reply:
@gregw the machinery can be customized, but without APIs. 
It's all relying on some magic "look at this attribute in the context" and the attribute names are based on classes that are not really meant to be exposed in this way, say for example `o.e.j.w.core.WebSocketComponents`.
I would not document this now, and if people really struggle, then add the docs later.